local-api/lib/formlabs_local_api/PreFormApi.py

```diff
@@ -65,6 +75,85 @@ def start_preform_server(pathToPreformServer=None, preform_port=44388):
                 preformApi.stop_preform_server()
                 print("PreForm server stopped.")
 
+    # TODO: reject_earlier_versions, reject_later_versions
+    @contextmanager
+    @staticmethod
+    def connect_to_preform_server(preform_port=44388):
+        """
+        Connect to an already-running PreForm server and yeild a PreFormApi client.
+        Usage:
+        ```
+        with PreFormApi.connect_to_preform_server() as preformApi:
+            preformApi.api.create_scene(...)
+        ```
+        """
+        preformApi = None
+        try:
+            process = PreFormApi.find_process_using_port(preform_port)
+            if process is None:
+                raise RuntimeError(f"No PreForm server found on port {preform_port}")
+            else:
+                PreFormApi.check_valid_server(preform_port)
+                yield preformApi
+                return
+        finally:
+            if preformApi:
+                preformApi.stop_preform_server()
+                print("PreForm server stopped.")
+
+
+    @contextmanager
+    @staticmethod
+    def start_or_connect_to_preform_server(pathToPreformServer=None, preform_port=44388):
+        """
+        connect_to_preform_server if a valid server is already running on this port, otherwise start a new server.
+
+        Usage:
+        ```
+        with PreFormApi.start_or_connect_to_preform_server() as preformApi:
+            preformApi.api.create_scene(...)
+        ```
+        """
+        # Check the preform port, attempt to connect to it
+        server_process = PreFormApi.find_process_using_port(preform_port)
+        if server_process is None:
+            preformApi = None
+            try:
+                preformApi = PreFormApi.start_preform_sync(pathToPreformServer, preform_port)
+                print("PreForm server ready")
+                yield preformApi
+                return
+            finally:
+                if (preformApi):
+                    preformApi.stop_preform_server()
+                    print("PreForm server stopped.")
```

Comment:
Can you reuse the above fn here?

Reply:
IMO with the yield, the cleanup, and possible None, reusing doesn't end up DRYing anything up.